gtest/src/manager.rs

```diff
@@ -398,9 +396,95 @@ impl ExtManager {
+        self.gas_tree
+            .create(dispatch.source(), dispatch.id(), gas_limit)
+            .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+        self.route_dispatch(dispatch)
+    }
+
+    pub(crate) fn route_dispatch(&mut self, dispatch: Dispatch) -> MessageId {
+        let stored_dispatch = dispatch.into_stored();
+        if self.is_user(&stored_dispatch.destination()) {
+            panic!("Program API only sends message to programs.")
+        }
+
+        let message_id = stored_dispatch.id();
+        self.dispatches.push_back(stored_dispatch);
+
+        message_id
+    }
+
+    // TODO #4120 Charge for task pool processing the gas from gas allowance
+    // TODO #4121
+    #[track_caller]
+    pub(crate) fn run_new_block(&mut self, allowance: Gas) -> BlockRunResult {
+        self.gas_allowance = allowance;
+        self.blocks_manager.next_block();
+        let new_block_bn = self.blocks_manager.get().height;
+
+        self.process_tasks(new_block_bn);
+        let total_processed = self.process_messages();
+
+        BlockRunResult {
+            block_info: self.blocks_manager.get(),
+            gas_allowance_spent: Gas(GAS_ALLOWANCE) - self.gas_allowance,
+            succeed: mem::take(&mut self.succeed),
+            failed: mem::take(&mut self.failed),
+            not_executed: mem::take(&mut self.not_executed),
+            total_processed,
+            log: mem::take(&mut self.log)
+                .into_iter()
+                .map(CoreLog::from)
+                .collect(),
+            gas_burned: mem::take(&mut self.gas_burned),
+        }
+    }
+
+    #[track_caller]
+    pub(crate) fn process_tasks(&mut self, bn: u32) {
+        self.process_delayed_dispatches(bn);
+        self.process_scheduled_wait_list(bn);
+    }
+
+    #[track_caller]
+    fn process_messages(&mut self) -> u32 {
+        self.messages_processing_enabled = true;
+
+        let mut total_processed = 0;
+        while self.messages_processing_enabled {
+            let dispatch = match self.dispatches.pop_front() {
+                Some(dispatch) => dispatch,
+                None => break,
+            };
+
+            let mut actors = self.actors.borrow_mut();
+            let (actor, balance) = actors
+                .get_mut(&dispatch.destination())
+                .expect("Somehow message queue contains message for user");
+            let balance = *balance;
+
+            if actor.is_dormant() {
+                drop(actors);
+                self.process_dormant(balance, dispatch);
+            } else if let Some((data, code)) = actor.get_executable_actor_data() {
+                drop(actors);
+                self.process_normal(balance, data, code, dispatch);
+            } else if let Some(mock) = actor.take_mock() {
+                drop(actors);
+                self.process_mock(mock, dispatch);
+            } else {
+                unreachable!();
```

Comment:
Manual drops seem very error-prone; maybe it could be improved with a `match`?

Reply:
discussed